relay-server/src/utils/managed_envelope.rs

```diff
@@ -129,6 +140,33 @@ impl ManagedEnvelope {
         self
     }
 
+    /// Retains only the items specified by the predicate.
+    ///
+    /// In other words, remove all elements where `f(&item)` returns `false`. This method operates
+    /// in place and preserves the order of the retained items.
+    pub fn retain_items<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut Item) -> RetainItem,
+    {
+        let mut outcomes = vec![];
+        let use_indexed = self.use_index_category();
+        self.envelope.retain_items(|item| match f(item) {
+            RetainItem::Keep => true,
+            RetainItem::Drop(outcome) => {
+                if let Some(category) = item.outcome_category(use_indexed) {
+                    outcomes.push((outcome, category, item.quantity()));
+                }
+
+                false
+            }
+            RetainItem::DropSilently => false,
+        });
+        for (outcome, category, quantity) in outcomes {
+            self.track_outcome(outcome, category, quantity);
+        }
```

Comment:
There's an `outcomes` vector that we populate during `retain_items`. Then, here, we track the outcomes of that vector. Is there any reason to not call `track_outcome` directly from inside `retain_items`?

Reply:
`track_outcome` borrows `self`, which is not allowed in this scope because `retain_items` borrows `self.envelope` as mutable. I also dislike the vector here though, open to any ideas how to get around the borrowing issue.

Reply:
I see. I think it's fine, it was more like a nit. Let's trust the borrow checker over me 😄.